apis/filesystem/src/test/java/org/jclouds/filesystem/integration/FilesystemBlobIntegrationTest.java

```diff
@@ -113,19 +121,59 @@ public void testSetBlobAccess() throws Exception {
    @Override
    protected void checkMPUParts(Blob blob, List<MultipartPart> partsList) {
       assertThat(blob.getMetadata().getETag()).endsWith(String.format("-%d\"", partsList.size()));
-      StringBuilder eTags = new StringBuilder();
+      Hasher eTagHasher = Hashing.md5().newHasher();
       for (MultipartPart part : partsList) {
-         eTags.append(part.partETag());
+         eTagHasher.putBytes(new BigInteger(part.partETag(), 16).toByteArray());
       }
       String expectedETag = new StringBuilder("\"")
-         .append(Hashing.md5().hashString(eTags.toString(), US_ASCII))
+         .append(eTagHasher.hash())
          .append("-")
          .append(partsList.size())
          .append("\"")
          .toString();
       assertThat(blob.getMetadata().getETag()).isEqualTo(expectedETag);
    }
 
+   @Test(groups = { "integration", "live" })
+   public void testMultipartUploadMultiplePartsKnownETag() throws Exception {
+      BlobStore blobStore = view.getBlobStore();
+      String container = getContainerName();
+      // Pre-computed ETag returned by AWS S3 for the MPU consisting of two 5MB parts filled with 'b'
+      String expectedETag = "\"84462a16f6a60478d50148808aa609c1-2\"";
+      int partSize = 5 * (int) Math.pow(2, 20);
+      try {
+         String name = "blob-name";
+         BlobBuilder blobBuilder = blobStore.blobBuilder(name);
+         Blob blob = blobBuilder.build();
+         MultipartUpload mpu = blobStore.initiateMultipartUpload(container, blob.getMetadata(), new PutOptions());
+
+         byte[] content = new byte[partSize];
+         Arrays.fill(content, (byte) 'b');
+         Payload payload = Payloads.newByteArrayPayload(content);
+
+         payload.getContentMetadata().setContentLength((long) partSize);
+
+         MultipartPart part1 = blobStore.uploadMultipartPart(mpu, 1, payload);
+         MultipartPart part2 = blobStore.uploadMultipartPart(mpu, 2, payload);
+
+         List<MultipartPart> parts = blobStore.listMultipartUpload(mpu);
+         assertThat(parts).hasSize(2);
+         assertThat(parts.get(0).partNumber()).isEqualTo(part1.partNumber());
+         assertThat(parts.get(0).partSize()).isEqualTo(part1.partSize());
+         assertThat(parts.get(0).partETag()).isEqualTo(part1.partETag());
+         assertThat(parts.get(1).partNumber()).isEqualTo(part2.partNumber());
+         assertThat(parts.get(1).partSize()).isEqualTo(part2.partSize());
+         assertThat(parts.get(1).partETag()).isEqualTo(part2.partETag());
+
+         blobStore.completeMultipartUpload(mpu, ImmutableList.of(part1, part2));
+
+         BlobMetadata newBlobMetadata = blobStore.blobMetadata(container, name);
+         assertThat(newBlobMetadata.getETag()).isEqualTo(expectedETag);
```

Comment:
[minor] Is there anything other than this final assertion (regarding the expected ETag) that we're trying to verify in this test?

Reply:
No, there isn't. I considered validating the content itself, but there are existing tests that do that. I can add the content validation if you see value to it.

Reply:
Sorry, I wasn't clear: I think it's fine to have *fewer* validations, so it's clear to the reader immediately what the "main" point of this test is. E.g. skip the assertions on number and size of parts, if that's not what we care about here?

Reply:
Gotcha! Will remove the extraneous checks.